File: sdk/typescript/src/bunqueue/triggers.ts
/**
 * Bunqueue Simple Mode — event triggers.
 * 1:1 port of src/client/bunqueue/triggers.ts: when a job completes/fails,
 * automatically create another job.
 */

import type { Job } from '../job.js';
import type { Queue } from '../queue.js';
import type { Worker } from '../worker.js';
import type { TriggerRule } from './types.js';

export class TriggerManager<T = unknown, R = unknown> {
  private readonly rules: TriggerRule<T>[] = [];
  private active = false;
  private readonly queue: Queue<T>;
  private readonly worker: Worker<T, R>;

  constructor(queue: Queue<T>, worker: Worker<T, R>) {
    this.queue = queue;
    this.worker = worker;
  }

  add(rule: TriggerRule<T>): void {
    this.rules.push(rule);
    this.ensureActive();
  }

  private ensureActive(): void {
    if (this.active) return;
    this.active = true;

    this.worker.on('completed', (job, result) => {
      this.fire('completed', job as Job<T>, result);
    });

    this.worker.on('failed', (job, error) => {
      this.fire('failed', job as Job<T>, error);
    });
  }

  private fire(event: 'completed' | 'failed', job: Job<T>, resultOrError: unknown): void {
    for (const rule of this.rules) {
      if (rule.on !== job.name) continue;
      if ((rule.event ?? 'completed') !== event) continue;
      if (rule.condition && !rule.condition(resultOrError, job)) continue;

      const data = rule.data(resultOrError, job);
      void this.queue.add(rule.create, data, rule.opts);
    }
  }
}

File: sdk/python/bunqueue/simple/triggers.py
"""Event triggers: when a job completes/fails, create another job (mirror of
triggers.ts)."""

from __future__ import annotations

from typing import Any, Dict, List

from ..queue import Queue
from ..worker import Worker


class TriggerManager:
    def __init__(self, queue: Queue, worker: Worker) -> None:
        self._queue = queue
        self._worker = worker
        self._rules: List[Dict[str, Any]] = []
        self._active = False

    def add(self, rule: Dict[str, Any]) -> None:
        """Rule keys: on, create, data(result, job), event ('completed'
        default | 'failed'), condition(result, job), opts (job options)."""
        self._rules.append(rule)
        self._ensure_active()

    def _ensure_active(self) -> None:
        if self._active:
            return
        self._active = True
        self._worker.on("completed", lambda job, result: self._fire("completed", job, result))
        self._worker.on("failed", lambda job, error: self._fire("failed", job, error))

    def _fire(self, event: str, job: Any, result_or_error: Any) -> None:
        for rule in self._rules:
            if rule["on"] != job.name:
                continue
            if rule.get("event", "completed") != event:
                continue
            condition = rule.get("condition")
            if condition and not condition(result_or_error, job):
                continue
            data = rule["data"](result_or_error, job)
            try:
                self._queue.add(rule["create"], data, **(rule.get("opts") or {}))
            except Exception:  # noqa: BLE001 - trigger add is fire-and-forget
                pass
